gtk4/src/bitset_iter.rs

```diff
@@ -0,0 +1,177 @@
+                Some((
+                    BitsetIter(iter.assume_init(), PhantomData),
+                    value.assume_init(),
+                ))
+            } else {
+                None
+            }
+        }
+    }
+
+    #[doc(alias = "gtk_bitset_iter_get_value")]
+    pub fn value(&self) -> u32 {
+        unsafe { ffi::gtk_bitset_iter_get_value(self.to_glib_none().0) }
+    }
+
+    #[doc(alias = "gtk_bitset_iter_is_valid")]
+    pub fn is_valid(&self) -> bool {
+        unsafe { from_glib(ffi::gtk_bitset_iter_is_valid(self.to_glib_none().0)) }
+    }
+
+    #[doc(alias = "gtk_bitset_iter_previous")]
+    pub fn previous(&mut self) -> Option<u32> {
+        unsafe {
+            let mut value = std::mem::MaybeUninit::uninit();
+
+            let found: bool = from_glib(ffi::gtk_bitset_iter_previous(
+                &mut self.0 as *mut _,
+                value.as_mut_ptr(),
+            ));
+            if found {
+                Some(value.assume_init())
+            } else {
+                None
+            }
+        }
+    }
+}
+
+impl<'a> Iterator for BitsetIter<'a> {
+    type Item = u32;
+
+    #[doc(alias = "gtk_bitset_iter_next")]
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
+            let mut value = std::mem::MaybeUninit::uninit();
+
+            let found: bool = from_glib(ffi::gtk_bitset_iter_next(
+                &mut self.0 as *mut _,
+                value.as_mut_ptr(),
+            ));
+            if found {
+                Some(value.assume_init())
+            } else {
+                None
+            }
+        }
+    }
+}
+
+#[doc(hidden)]
+impl<'a> ToGlibPtr<'a, *const ffi::GtkBitsetIter> for BitsetIter<'a> {
+    type Storage = &'a Self;
+
+    #[inline]
+    fn to_glib_none(&'a self) -> Stash<*const ffi::GtkBitsetIter, Self> {
+        Stash(&self.0 as *const ffi::GtkBitsetIter, self)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_bitset_iter() {
+        crate::init().unwrap();
+
+        let set = Bitset::new_range(0, 20);
+        let (mut iter, init_value) = BitsetIter::init_first(&set).unwrap();
+        assert_eq!(init_value, 0);
+        assert_eq!(iter.next(), Some(1));
+        assert_eq!(iter.previous(), Some(0));
+
+        let set2 = Bitset::new_range(0, 3);
+        let (mut iter, init_val) = BitsetIter::init_last(&set2).unwrap();
+        assert_eq!(init_val, 2);
+        assert_eq!(iter.previous(), Some(1));
+        assert_eq!(iter.previous(), Some(0));
+        assert_eq!(iter.previous(), None);
+        assert!(!iter.is_valid());
+        assert_eq!(iter.next(), Some(1));
+        assert!(iter.is_valid());
+    }
+
+    #[test]
+    fn test_bitset_iter_dropped_bitset() {
+        crate::init().unwrap();
+
+        let set = Bitset::new_range(0, 20);
+        let (mut iter, init_value) = BitsetIter::init_first(&set).unwrap();
+        unsafe {
+            ffi::gtk_bitset_unref(set.to_glib_none().0);
```

Comment:
Well of course this doesn't work. You're using `unsafe` code here to force dropping the bitset so the compiler can't protect you :)

If you'd try using `drop(set)` instead, the compiler would prevent you from breaking things.

Reply:
aha interesting
```
error[E0505]: cannot move out of `set` because it is borrowed
   --> gtk4/src/bitset_iter.rs:169:14
    |
168 |         let (mut iter, init_value) = BitsetIter::init_first(&set).unwrap();
    |                                                             ---- borrow of `set` occurs here
169 |         drop(set);
    |              ^^^ move out of `set` occurs here
170 |         assert_eq!(init_value, 0);
171 |         assert_eq!(iter.next(), Some(1));
    |                    ---- borrow later used here
```
Nice! thanks for the tip, learned something new :)